src/pyqasm/visitor.py

```diff
@@ -559,7 +562,66 @@ def _check_variable_cast_type(
                 span=statement.span,
             )
 
-    def _visit_measurement(  # pylint: disable=too-many-locals
+    def _get_pyqasm_device_qubit_index(self, reg: str, idx: int) -> int:
+        """
+        Returns qubit index in __PYQASM_QUBITS__ ordering for given quantum register and index.
+
+        Args:
+            reg (str): The name of the quantum register.
+            idx (int): The index within the quantum register.
+
+        Returns:
+            int: The global __PYQASM_QUBITS__ index corresponding to reg[idx].
+
+        Raises:
+            IndexError: If the index `idx` is out of range for the register.
+        """
+        _offsets: dict[str, int] = {}
+        _offset = 0
+        for name, n_qubits in self._global_qreg_size_map.items():
+            _offsets[name] = _offset
+            _offset += n_qubits
+        _n_qubits = self._global_qreg_size_map[reg]
+        if not 0 <= idx < _n_qubits:
+            raise IndexError(f"{reg}[{idx}] out of range (0..{_n_qubits-1})")
+        return _offsets[reg] + idx
+
+    def _qubit_register_consolidation(self, unrolled_stmts: list) -> list[qasm3_ast.Statement]:
+        """
+        Consolidate all quantum registers into a single register '__PYQASM_QUBITS__'.
+
+        Args:
+            unrolled_stmts (list): The list of statements to process and modify in-place.
+
+        Raises:
+            ValidationError: If the total number of qubits exceeds the available device qubits,
+                             or if the reserved register '__PYQASM_QUBITS__' is already declared
+                             in the original QASM program.
+        """
+        total_qubits = sum(self._global_qreg_size_map.values())
+        if total_qubits > self._device_qubits:
+            raise_qasm3_error(
+                f"Total qubits '({total_qubits})' exceed device qubits '({self._device_qubits})'."
+            )
+
+        if "__PYQASM_QUBITS__" in self._global_qreg_size_map:
+            raise_qasm3_error(
+                "Original QASM program already declares reserved register '__PYQASM_QUBITS__'."
+            )
+
+        removable_statements = []
+        for stmt in unrolled_stmts:
+            if isinstance(stmt, qasm3_ast.QubitDeclaration):
+                removable_statements.append(stmt)
+        for r_stmt in removable_statements:
```

Comment:
It is better to construct a new list of `valid_statements` which do not contain the original `QubitDeclaration`s. You can just append any statement which is not of type `QubitDeclaration`. 

The reason is that `.remove(...)` method works in `O(n)` time, `n` being the size of our list. For a large number of qubit declarations, say `k`, this will scale as `O(n*k)`.